[scudo] Add specific die functions for linux specific failures.

cferris1000 opened the pull request:
While running into failures on unmap calls, it becomes difficult to figure out what's wrong. Break the dieOnMapUnmapError into specific versions for map, unmap, and then one for mprotect.

Also, put these in a common linux space so that all linux derived code can reuse this code.

cferris1000 committed:
[scudo] Add specific Report functions for linux specific failures.

While running into failures on unmap calls, it becomes difficult
to figure out what's wrong. Break the dieOnMapUnmapError into
specific Report versions for map, unmap, and then one for mprotect.

Also, put these in a common report linux file so that all linux derived
code can reuse this code.

Finally, add a ReportRawError that is used by all of the code so that
it's all in the same space.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/linux_common.cpp:
```diff
@@ -0,0 +1,65 @@
+
+namespace scudo {
+
+void NORETURN die() { abort(); }
```
nit: If this is only used here, I may mark it as static and put it in the global namespace.

cferris1000 replied:
It's used by the report code. I moved it back into the linux.cpp and trusty.cpp files since it feels like it better belongs there. Even though the die on trusty and linux are the same, it's possible they could be changed so leave it

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/CMakeLists.txt:
```diff
@@ -108,6 +108,7 @@ set(SCUDO_SOURCES
   flags.cpp
   fuchsia.cpp
   linux.cpp
+  linux_common.cpp
```
Given that we have `report.h` to report platform independent errors, I'm thinking that we may want the naming to be something like `report_linux_common.h` (and linux.cpp will be deprecated)

cferris1000 replied:
I decided to change this to report_linux since having common in the name seems redundant.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/linux_common.h:
```diff
@@ -0,0 +1,34 @@
+// Internal unmap fatal error. This must not call map().
+void NORETURN dieOnUnmapError(uptr Addr, uptr Size);
+
+// Internal protect fatal error. This must not call map().
```
Given that `map()` will be deprecated, maybe we can remove (or update) `This must not call map()`?

cferris1000 replied:
I changed the comments a bit to convey what's going on a bit better.

ChiaHungDuan reviewed (commented):
Only very few coding style suggestions

cferris1000 committed:
Update for comments.

Move the dieOnXXX routines to reportXXX. Mv linux_common to report_linux.

Add a new reportRawError function that doesn't use ScopedString.

cferris1000 committed:
Fix names in report_linux.{cpp,h}.

Move the die definitions in trusty.cpp and linux.cpp to their original
location to avoid unnecessary churn in the files.

ChiaHungDuan reviewed (approved):
Looks nice!

cferris1000 committed:
Run clang-format.

fabio-d commented on compiler-rt/lib/scudo/standalone/CMakeLists.txt:
```diff
@@ -113,6 +113,7 @@ set(SCUDO_SOURCES
   mem_map_linux.cpp
   release.cpp
   report.cpp
+  report_linux.cpp
```
Should the corresponding `.h` file be listed too in the section above?

cferris1000 replied:
Yep, thanks for spotting that. I've fixed it now.

cferris1000 committed:
Adding report_linux.h to list of headers.

cferris1000 committed:
Add TRUSTY for report_linux.{cpp,h}.

cferris1000 commented on compiler-rt/lib/scudo/standalone/trusty.cpp:
```diff
@@ -12,6 +12,7 @@
 
 #include "common.h"
 #include "mutex.h"
+#include "report_linux.h"
```
I added a check for trusty too. I misread the platform.h as setting SCUDO_LINUX when trusty is defined.

cferris1000 commented:
Fixed the comment on the first message to actually reflect what was done.

ChiaHungDuan commented:
Still LGTM

fabio-d reviewed (approved):
Thanks!
